### PR TITLE
pinentry-mac: 0.9.4 -> 1.1.1.1

### PR DESCRIPTION
**Pull request by midchildan**
###### Description of changes

Upgrades pinentry-mac. The upstream repository moved from https://github.com/GPGTools/pinentry-mac to https://github.com/GPGTools/pinentry back in [2018](https://github.com/GPGTools/pinentry/commit/d35b8c73e55a5f4151e008c2d6ceb04051d8e155), but seemingly went unnoticed.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

The result of running `echo GETPIN | result/Applications/pinentry-mac.app/Contents/MacOS/pinentry-mac`:
Screen Shot 2022-06-04 at 4 04 38 PM

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
This shouldn't go to staging but straight to master

**Comment by midchildan** [comment]
I switched the target branch back to `master` and rebased it against the latest commit.

**Comment by SuperSandro2000** [comment]
@ofborg build pinentry_mac

**Comment by midchildan** [comment]
> /usr/bin/ibtool --compile Main.nib Main.xib
> xcode-select: error: no developer tools were found at '/Applications/Xcode.app', and no install could be requested (perhaps no UI is present), please install manually from 'developer.apple.com'.
> make[2]: *** [Makefile:741: Main.nib] Error 1

Hmm, so I guess this is unbuildable on Hydra. I could think of 3 options to deal with this problem:

1. Generate the nib files locally and include them in Nixpkgs.
2. Set `hydraPlatforms` to `[]`. Might be problematic because it seems 100 ~ 500 packages depend on pinentry-mac.
3. Stick with pinentry-mac 0.9.4.

My preference is option 1.

**Comment by SuperSandro2000** [comment]
or can we use the tools in nixpkgs?

**Comment by midchildan** [comment]
There's no complete open source ibtool replacement that I'm aware of. There's xib2nib, but it doesn't work for pinentry_mac:

```
       > Unhandled attribute:/document/objects/window/point:canvasLocation:x:71
       > Unhandled attribute:/document/objects/window/point:canvasLocation:y:-49
       > Unhandled node:/document/objects/customFormatter:(null)
       > Unhandled attribute:/document/objects/customFormatter:(null):customClass:PassphraseLengthFormatter
       > Unhandled node:/document/resources:(null)
       > Unhandled node:/document/resources/image:(null)
       > Unhandled attribute:/document/resources/image:(null):name:Icon
       > Unhandled attribute:/document/resources/image:(null):width:512
       > Unhandled attribute:/document/resources/image:(null):height:512
       > /nix/store/cm4394z3j2kmcbn3ajhyv2jmb4k3jfmj-stdenv-darwin/setup: line 96: 75124 Segmentation fault: 11  xib2nib Pinentry.xib Pinentry.nib
```

**Comment by midchildan** [comment]
I updated this PR with this approach:

> Generate the nib files locally and include them in Nixpkgs.

File: pkgs/tools/security/pinentry/mac/.editorconfig [hunk 1]
```diff
@@ -0,0 +1 @@
+root = true
```
Comment (midchildan) [inline pkgs/tools/security/pinentry/mac/.editorconfig:1]: This is here to stop editorconfig from throwing errors on binary nib files.

Comment (SuperSandro2000) [reply pkgs/tools/security/pinentry/mac/.editorconfig:1]: Please move this into the top-level file

Comment (midchildan) [reply pkgs/tools/security/pinentry/mac/.editorconfig]: Done.

**Review by Luflosi — APPROVED** [review]
LGTM.

**Comment by SuperSandro2000** [comment]
I don't really want to check in the binary files into nixpkgs...

**Comment by midchildan** [comment]
Agreed, but there don't seem to be any other options:

* Removing pinentry-mac from Hydra is problematic because it's popular and affects too many existing packages in Nixpkgs
* Holding off updates indefinitely is especially problematic for security-related software

This means we have to generate the nib files somehow, but that's not practically possible without using ibtools from XCode, which in turn can't be done from within the sandbox.

It's not great, but at least in the case of pinentry-mac the binaries are only 70 KB total. Judging by the release history, updates are infrequent too.

**Comment by Luflosi** [comment]
Would it be a good idea to host these files elsewhere and then only store the URL and hash in Nixpkgs?

**Comment by midchildan** [comment]
That requires someone to maintain the files independently of Nixpkgs, which would limit who would be able to perform updates. It's doable, but I work it might lead to updates being hold off for a long time.

**Comment by ylh** [comment]
Yes, checking in binaries feels a little gross, but it's worth noting that `Main.xib` and `Pinentry.xib` have not changed in years, and going by the commit history, they're unlikely to change that often going forward. If by some run of bad luck they do, *and* the corresponding version is security-critical, it's always possible to mark as insecure while we scramble to generate fresh nibs.

**Review by toonn — APPROVED** [review]
LGTM

Built with `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 176203 -p pinentry_mac"`

And tested with: `echo GETPIN | results/pinentry_mac/Applications/pinentry-mac.app/Contents/MacOS/pinentry-mac`

The first time I ran this I got:

```
bash: cho: command not found            
OK Pleased to meet you
```

Not sure what that was about but it did not reoccur.

Carrying binaries in Nixpkgs isn't great but until something like xib2nib becomes capable of generating the nibs for this I don't really see a good alternative.

**Pull request merged**